rules/DowngradePhp70/Rector/Expr/DowngradeUnnecessarilyParenthesizedExpressionRector.php

```diff
@@ -0,0 +1,90 @@
+
+namespace Rector\DowngradePhp70\Rector\Expr;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr;
+use PhpParser\Node\Expr\ArrayDimFetch;
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\PropertyFetch;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\Expr\StaticPropertyFetch;
+use Rector\Core\Rector\AbstractRector;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://wiki.php.net/rfc/uniform_variable_syntax
+ *
+ * @see \Rector\Tests\DowngradePhp70\Rector\Expr\DowngradeUnnecessarilyParenthesizedExpressionRector\DowngradeUnnecessarilyParenthesizedExpressionRectorTest
+ */
+final class DowngradeUnnecessarilyParenthesizedExpressionRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Remove parentheses around expressions where they are not necessary thanks to Uniform variable syntax RFC.',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+($f)['foo'];
+($f)->foo;
+($f)->foo();
+($f)::$foo;
+($f)::foo();
+($f)();
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+$f['foo'];
+$f->foo;
+$f->foo();
+$f::$foo;
+$f::foo();
+$f();
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        // TODO: Make PHPStan allow Expr namespace.
+        $dummyClassNameUse = str_contains(Expr::class, '\\');
+        assert($dummyClassNameUse);
+        return [
+            ArrayDimFetch::class,
+            PropertyFetch::class,
+            MethodCall::class,
+            StaticPropertyFetch::class,
+            StaticCall::class,
+            FuncCall::class,
+        ];
+    }
+
+    /**
+     * @param ArrayDimFetch|PropertyFetch|MethodCall|StaticPropertyFetch|StaticCall|FuncCall $node
+     */
+    public function refactor(Node $node): ?Expr
+    {
+        // Re-create the node to strip redundant parentheses around the expression
+        // since PHP 5 cannot deal with them.
+        // In some cases parens are necessary so they will remain even after this rule is applied
+        // but other rules should handle that.
+        return match (true) {
+            $node instanceof ArrayDimFetch => new ArrayDimFetch($node->var, $node->dim),
+            $node instanceof PropertyFetch => new PropertyFetch($node->var, $node->name),
+            $node instanceof MethodCall => new MethodCall($node->var, $node->name, $node->args),
+            $node instanceof StaticPropertyFetch => new StaticPropertyFetch($node->class, $node->name),
+            $node instanceof StaticCall => new StaticCall($node->class, $node->name, $node->args),
+            default /* $node instanceof FuncCall */ => new FuncCall($node->name, $node->args),
```

Comment:
Another PHPStan meh.

Reply:
PHPStan is right here, as this technically works, but `match (true)` magic is not suitable for type matching. 

Here would be better collector pattern as we have e.g. in `NodeNameResolver`.

Reply:
Could you be more specific? I do not see anything applicable in `NodeNameResolver` and collector pattern does not return any relevant results.

Reply:
The `NodeNameResolver` accepts array of `array $nodeNameResolvers = []` that each resolve own types based on instanceof.
Before we had this huge single class that contained everything. This now allows to focus on one node at a time.